File: backend/app/schemas/assignment.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from app.enums.allocation_type import AssignmentAllocationType
from app.schemas.base import JsonModel


class AssignmentCreate(JsonModel):
    project_id: UUID
    phase_id: UUID
    employee_id: UUID
    allocation_type: AssignmentAllocationType
    allocated_hours: int | None = None
    allocation_percentage: int | None = None
    hourly_rate_override: Decimal | None = None
    start_date: date
    end_date: date


class AssignmentUpdate(JsonModel):
    allocation_type: AssignmentAllocationType | None = None
    allocated_hours: int | None = None
    allocation_percentage: int | None = None
    hourly_rate_override: Decimal | None = None
    start_date: date | None = None
    end_date: date | None = None


class AssignmentResponse(JsonModel):
    id: UUID
    project_id: UUID
    phase_id: UUID
    employee_id: UUID
    allocation_type: AssignmentAllocationType
    allocated_hours: int | None
    allocation_percentage: int | None
    hourly_rate_override: Decimal | None
    start_date: date
    end_date: date
    created_at: datetime
    updated_at: datetime

File: client/src/types/assignment.ts
export const AssignmentAllocationType = {
  HOURS: "hours",
  PERCENTAGE: "percentage",
} as const;

export type AssignmentAllocationType =
  (typeof AssignmentAllocationType)[keyof typeof AssignmentAllocationType];

export interface Assignment {
  id: string;
  projectId: string;
  phaseId: string;
  employeeId: string;
  allocationType: AssignmentAllocationType;
  allocatedHours: number | null;
  allocationPercentage: number | null;
  hourlyRateOverride: number | null;
  startDate: string;
  endDate: string;
  createdAt: string;
  updatedAt: string;
}

export interface AssignmentWithDetails extends Assignment {
  employeeName: string;
  projectName: string;
  phaseName: string;
}

export interface AssignmentCreate {
  projectId: string;
  phaseId: string;
  employeeId: string;
  allocationType: AssignmentAllocationType;
  allocatedHours?: number;
  allocationPercentage?: number;
  hourlyRateOverride?: number;
  startDate: string;
  endDate: string;
}
